### PR TITLE
Improve logging

### LINKED ISSUE #147: Improve logging
Currently only the default meinheld server logging is done (and initial predeployed accounts logging).

1. The server logging could be optional (to be handled with gunicorn).
2. More informative logging could be done on each:
    - transaction (deploy, declare, invoke)
    - emitted event
    - time manipulation
    - minting
3. Users could opt for more or less verbose logging (`-v, --verbose`)
4. Log all predeployed contract information (e.g. fee token, UDC)

### PR DESCRIPTION
**Pull request by tonypony220**
## Usage related changes
Closes #147 

- added `--verbose` arg
- added `--hide-server-logs` arg to hide access logging
- added `--hide-predeployed-contracts` arg to substitute `--hide-predeployed-accounts`
- added warning to not use it

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/lint.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Updated the tests
- [x] All tests are passing - `./scripts/test.sh`

### CODE REVIEW

**Comment by FabijanC** [comment]
Thanks for the PR! I'll take a look next week.

**Comment by FabijanC** [comment]
Just to let you know that I didn't forget about this, but currently adapting our tools to the latest cairo-lang/starknet/compiler versions is the highest priority.

**Comment by FabijanC** [comment]
Btw @tonypony220 I see some of the checkboxes in the PR description aren't checked. Do you plan to address them or do you think they are not necessary?

**Comment by tonypony220** [comment]
Of course! I've done fresh rebase and run tests.

File: starknet_devnet/util.py [hunk 1]
```diff
@@ -263,3 +264,62 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            if isinstance(exc_val, StarkException):
+                error_message = {
+                    "message": exc_val.message,
+                    "code": str(exc_val.code),
+                    "context": self,
+                }
+            else:
+                error_message = {"message": str(exc_val), "context": self}
+            logger.error(
+                "%s failed: %s", self.get_context_name(), pprint.pformat(error_message)
```
Comment (FabijanC) [inline starknet_devnet/util.py:307]: What's the benefit of using `pprint`? Asking because I'm not experienced with it.

File: starknet_devnet/util.py [hunk 2]
```diff
@@ -263,3 +264,62 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+
+logger = logging.getLogger("gunicorn.error")
+
+
+class LogContext(dict):
+    """Context manager for logging transactions accumulating info while executing transaction"""
+
+    def get_context_name(self) -> str:
+        """Getting name of current context to put as first line in log entry"""
+        return getattr(self, "name", None) or "Transaction"
```
Comment (FabijanC) [inline starknet_devnet/util.py:276]: Shouldn't it be `getattr(self, "name", "Transaction")`? Or even better, this class could have a property `name` 
 defined just before this method (initially set to `None`).

File: starknet_devnet/util.py [hunk 3]
```diff
@@ -263,3 +264,62 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+
+    def set_context_name(self, name):
```
Comment (FabijanC) [inline starknet_devnet/util.py:278]: Set the return value? And annotate parameter type?
[Useful link](https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class)

File: starknet_devnet/util.py [hunk 4]
```diff
@@ -263,3 +264,62 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+        """Setting name of current context to put as first line in log entry"""
+        setattr(self, "name", name)
+        return self
+
+    def cust_fees_to_int(self):
+        """Cust to human readable type"""
+        for key in ["actual_fee", "max_fee"]:
+            if self.get(key):
+                try:
```
Comment (FabijanC) [inline starknet_devnet/util.py:287]: Looks hacky. If not rewritten, it could at least be better documented.

File: starknet_devnet/util.py [hunk 5]
```diff
@@ -263,3 +264,62 @@ def __exit__(self, exc_type, exc_val, exc_tb):
...
+            else:
+                error_message = {"message": str(exc_val), "context": self}
+            logger.error(
+                "%s failed: %s", self.get_context_name(), pprint.pformat(error_message)
+            )
+            return
+        logger.info("%s completed: %s", self.get_context_name(), pprint.pformat(self))
+
+
+def extract_transaction_info_to_log(transaction: dict) -> dict:
```
Comment (FabijanC) [inline starknet_devnet/util.py:313]: Why is this necessary? Why not just print the object? It is called the "verbose" mode? Can the output be huge?

File: starknet_devnet/udc.py [hunk 6]
```diff
@@ -44,3 +46,10 @@ def contract_class(self) -> CompiledClassBase:
 
     async def _mimic_constructor(self):
         pass
+
+    def _print(self):
+        """stdout UDC"""
```
Comment (FabijanC) [inline starknet_devnet/udc.py:51]: If the parent `_print` has a docstring, the children don't need it (especially if it's not really informative). But then the parent can have a more verbose docstring.

File: starknet_devnet/starknet_wrapper.py [hunk 7]
```diff
@@ -416,7 +418,7 @@ def _update_block_number(self):
         )
         return block_info
 
-    def __get_transaction_handler(self, external_tx=None):
+    def __get_transaction_handler(self, external_tx=None, context: LogContext = None):
```
Comment (FabijanC) [inline starknet_devnet/starknet_wrapper.py:421]: Here and in other places, instead of passing the context object around, couldn't we modify a global object?
[Related SO question](https://stackoverflow.com/questions/50556584/should-logger-be-an-argument-or-a-global-variable)

**Review by FabijanC — CHANGES_REQUESTED** [review]
Why didn't you check the linter checkbox?

File: starknet_devnet/starknet_wrapper.py (hunk 7 above)
Comment (tonypony220) [reply starknet_devnet/starknet_wrapper.py:421]: I had the thought that someday the app would be used asynchronously, making it difficult to comprehend the sequence of log entries. To address this, I created a context that keeps track of a single transaction and outputs information when it is fully completed. 
It seems I'm wrong, I will rewrite it to log the progress at each stage.

File: page/docs/guide/run.md [hunk 8]
```diff
@@ -18,7 +18,8 @@ Run a local instance of Starknet Devnet
 
 optional arguments:
   -h, --help            show this help message and exit
-  -v, --version         Print the version
+  -V, --version         Print the version
```
[the remark this answers is not in the document]
Comment (tonypony220) [reply page/docs/guide/run.md]: done

File: starknet_devnet/account.py [hunk 9]
```diff
@@ -58,3 +58,6 @@ async def _mimic_constructor(self):
         )
 
         await set_balance(starknet.state, self.address, self.initial_balance)
+
+    def _print(self):
+        """Predeployed Accounts logged in Accounts Wrapper"""
```
[the remark this answers is not in the document]
Comment (tonypony220) [reply starknet_devnet/account.py]: Made public

File: starknet_devnet/util.py (hunk 4 above)
Comment (tonypony220) [reply starknet_devnet/util.py]: removed this class

File: starknet_devnet/util.py (hunk 5 above)
Comment (tonypony220) [reply starknet_devnet/util.py]: Sure, made it more verbose

File: starknet_devnet/udc.py (hunk 6 above)
Comment (tonypony220) [reply starknet_devnet/udc.py]: Thx!

File: starknet_devnet/util.py (hunk 1 above)
Comment (tonypony220) [reply starknet_devnet/util.py]: It prints json with indentation (like jq cli tool) and it's very readable

File: starknet_devnet/util.py (hunk 2 above)
Comment (tonypony220) [reply starknet_devnet/util.py]: removed cls

File: starknet_devnet/util.py (hunk 5 above)
Comment (FabijanC) [reply starknet_devnet/util.py]: I meant my last question (Can the output be huge?) more like: did you reduce the number of printed lines because we are risking the output to be huge and unreadable?

Comment (tonypony220) [reply starknet_devnet/util.py]: I decided to reduce it because I believed that the user wouldn't require such detailed information, only concise "INFO" entries, which could have been extended with context. But in the end, we don't need to keep it at a small size.

With the full verbose mode, it remains readable, except for some contracts that mainly consist of numbers, and I'm unsure if the user would be able to understand them.

Here an example of size deploy transaction with full output: 
```
[2023-06-08 10:06:22 +0000] [5354] [INFO] add_transaction request: {'class_hash': '0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f',
 'constructor_calldata': ['0x4024999b9574cb7623679ce049a609db62a95098982c5b28ac61abdebd1c82b'],
 'contract_address_salt': '0x99',
 'max_fee': '0xde0b6b3a7640000',
 'nonce': '0x0',
 'signature': ['0x15dc17a076cea6ec2355cde875c7f46b0ff5f312b418afed547f418622b9298',
               '0x246e0207bb454a967abc0b336f6705d7a521017a119d05953cbc25d2e04d4e1'],
 'type': 'DEPLOY_ACCOUNT',
 'version': '0x1'}
[2023-06-08 10:06:23 +0000] [5354] [INFO] transaction execution info: {'actual_fee': '0x64a7168300',
 'actual_resources': {'ecdsa_builtin': 1,
                      'l1_gas_usage': 4284,
                      'n_steps': 3818,
                      'pedersen_builtin': 23,
                      'range_check_builtin': 83},
 'call_info': {'accessed_storage_keys': ['0x1379ac0624b939ceb9dede92211d7db5ee174fe28be72245b0a1a2abd81c98f'],
               'call_type': 'CALL',
               'calldata': [1813293060336465068570802339647185657676088914728234993269889516456358561835],
               'caller_address': 0,
               'class_hash': '0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f',
               'code_address': None,
               'contract_address': 1065893459572464416143945688077233223288051643985551879776468107551438372322,
               'entry_point_selector': 1159040026212278395030414237414753050475174923702621880048416706425641521556,
               'entry_point_type': 'CONSTRUCTOR',
               'events': [],
               'execution_resources': {'builtin_instance_counter': {},
                                       'n_memory_holes': 0,
                                       'n_steps': 48},
               'failure_flag': '0x0',
               'gas_consumed': '0x0',
               'internal_calls': [],
               'l2_to_l1_messages': [],
               'retdata': [],
               'storage_read_values': [0]},
 'fee_transfer_info': {'accessed_storage_keys': ['0x249c2e4c4e43fd9200796b7abc7f94d3840e2cb6af5cbfba2526ff1c8c28729',
                                                 '0x249c2e4c4e43fd9200796b7abc7f94d3840e2cb6af5cbfba2526ff1c8c2872a',
                                                 '0x43d06bc782c9584c9b9cae590f47bc51fc5ddcbb23d2c0cbe553b86c8f0e040',
                                                 '0x43d06bc782c9584c9b9cae590f47bc51fc5ddcbb23d2c0cbe553b86c8f0e041'],
                       'call_type': 'CALL',
                       'calldata': [1815204096841571598366742751903307880575707433299601255915223373127742545712,
                                    432300000000,
                                    0],
                       'caller_address': 1065893459572464416143945688077233223288051643985551879776468107551438372322,
                       'class_hash': '0x6a22bf63c7bc07effa39a25dfbd21523d211db0100a0afd054d172b81840eaf',
                       'code_address': None,
                       'contract_address': 2087021424722619777119509474943472645767659996348769578120564519014510906823,
                       'entry_point_selector': 232670485425082704932579856502088130646006032362877466777181098476241604910,
                       'entry_point_type': 'EXTERNAL',
                       'events': [{'data': ['0x25b4639c321f3538c69537502f0379a23d6c91d6bf0f9dfe79fabfc3da97de2',
                                            '0x4035e7e01f09b8bc746a549fdf1bafa8d975bdbc53d25646635d0ab58425f30',
                                            '0x64a7168300',
                                            '0x0'],
                                   'keys': ['0x99cd8bde557814842a3121e8ddfd433a539b8c9f14bf31ebf108d12e6196e9'],
                                   'order': 0}],
                       'execution_resources': {'builtin_instance_counter': {'pedersen_builtin': 4,
                                                                            'range_check_builtin': 29},
                                               'n_memory_holes': 42,
                                               'n_steps': 578},
                       'failure_flag': '0x0',
                       'gas_consumed': '0x0',
                       'internal_calls': [],
                       'l2_to_l1_messages': [],
                       'retdata': [1],
                       'storage_read_values': [1000000000000000000,
                                               0,
                                               1000000000000000000,
                                               0,
                                               495000000000,
                                               0,
                                               495000000000,
                                               0]},
 'tx_type': 'DEPLOY_ACCOUNT',
 'validate_info': {'accessed_storage_keys': ['0x1379ac0624b939ceb9dede92211d7db5ee174fe28be72245b0a1a2abd81c98f'],
                   'call_type': 'CALL',
                   'calldata': [2177626953842241688342849477958144159584215338353441986251810208376165874783,
                                153,
                                1813293060336465068570802339647185657676088914728234993269889516456358561835],
                   'caller_address': 0,
                   'class_hash': '0x4d07e40e93398ed3c76981e72dd1fd22557a78ce36c0515f679e27f0bb5bc5f',
                   'code_address': None,
                   'contract_address': 1065893459572464416143945688077233223288051643985551879776468107551438372322,
                   'entry_point_selector': 1554466106298962091002569854891683800203193677547440645928814916929210362005,
                   'entry_point_type': 'EXTERNAL',
                   'events': [],
                   'execution_resources': {'builtin_instance_counter': {'ecdsa_builtin': 1},
                                           'n_memory_holes': 0,
                                           'n_steps': 80},
                   'failure_flag': '0x0',
                   'gas_consumed': '0x0',
                   'internal_calls': [],
                   'l2_to_l1_messages': [],
                   'retdata': [],
                   'storage_read_values': [1813293060336465068570802339647185657676088914728234993269889516456358561835]}}
[2023-06-08 10:06:23 +0000] [5354] [INFO] transaction receipt: {'actual_fee': '0x64a7168300',
 'block_hash': '0x11c4ef8d04df9d839e90f13f43b8f8b87be1057154674cfe875b6673c878845',
 'block_number': 1,
 'events': [{'data': ['0x25b4639c321f3538c69537502f0379a23d6c91d6bf0f9dfe79fabfc3da97de2',
                      '0x4035e7e01f09b8bc746a549fdf1bafa8d975bdbc53d25646635d0ab58425f30',
                      '0x64a7168300',
                      '0x0'],
             'from_address': '0x49d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7',
             'keys': ['0x99cd8bde557814842a3121e8ddfd433a539b8c9f14bf31ebf108d12e6196e9']}],
 'execution_resources': {'builtin_instance_counter': {},
                         'n_memory_holes': 0,
                         'n_steps': 48},
 'l2_to_l1_messages': [],
 'status': 'ACCEPTED_ON_L2',
 'transaction_hash': '0x58a5b426cbda66d208fc300d74e6f371f414feb27f7ff0897efba04b563a5fb',
 'transaction_index': 0}
[2023-06-08 10:06:23 +0000] [5354] [INFO] add_transaction response: {'address': '0x025b4639c321f3538c69537502f0379a23d6c91d6bf0f9dfe79fabfc3da97de2',
 'code': 'TRANSACTION_RECEIVED',
 'transaction_hash': '0x58a5b426cbda66d208fc300d74e6f371f414feb27f7ff0897efba04b563a5fb'}
[2023-06-08 10:06:23 +0000] [5354] [INFO] 127.0.0.1 - - [08/Jun/2023:10:06:23 +0000] "POST /gateway/add_transaction HTTP/1.1" 200 198 "-" "python-requests/2.28.2"
```

**Comment by tonypony220** [comment]
Should I make all conversations resolved?

**Comment by FabijanC** [comment]
Just now I realized that this PR is still introducing a breaking change:
> changes args `--hide-predeployed-accounts` to `--hide-predeployed-contracts`

Do you have an opinion on that? I really wouldn't want to bump to 0.6.0 just because of this

File: starknet_devnet/blueprints/gateway.py [hunk 10]
```diff
@@ -28,9 +29,10 @@ async def add_transaction():
     }
 
     if tx_type == TransactionType.DECLARE:
-        contract_class_hash, transaction_hash = await state.starknet_wrapper.declare(
-            transaction
-        )
+        (
```
Comment (FabijanC) [inline starknet_devnet/blueprints/gateway.py:32]: How was this change induced? Can you format the code?

File: log [hunk 11]
```diff
@@ -0,0 +1,311 @@
+============================= test session starts ==============================
```
Comment (FabijanC) [inline log:1]: Thanks for putting an example here. I assume we should delete it before merging?

File: starknet_devnet/blueprints/rpc/transactions.py [hunk 12]
```diff
@@ -123,7 +123,6 @@ async def add_declare_transaction(
     """
     if int(declare_transaction["version"], 0) == LEGACY_TX_VERSION:
         raise RpcError.from_spec_name("INVALID_CONTRACT_CLASS")
-
```
Comment (FabijanC) [inline starknet_devnet/blueprints/rpc/transactions.py:126]: I notice in a lot of places you delete some of the empty lines. What is your motivation for that?

File: starknet_devnet/devnet_config.py [hunk 13]
```diff
@@ -317,9 +327,9 @@ def parse_args(raw_args: List[str]):
         help="Specify the seed for randomness of accounts to be predeployed",
     )
     parser.add_argument(
-        "--hide-predeployed-accounts",
```
Comment (FabijanC) [inline starknet_devnet/devnet_config.py]: Perhaps instead of introducing a breaking change, we could keep the `--hide-predeployed-accounts` flag, but just have it print a deprecation notice and notify the user to use the other flag. Or this `parser.add_argument(...` could accept two entry points (`--hide-predeployed-accounts` and `--hide-predeployed-contracts`)

File: starknet_devnet/devnet_config.py [hunk 14]
```diff
@@ -264,6 +264,16 @@ def parse_args(raw_args: List[str]):
...
     )
+    parser.add_argument(
+        "--verbose",
+        action="store_true",
+        help="Show more verbose logs",
+    )
+    parser.add_argument(
+        "--hide-logs",
+        action="store_true",
+        help="Hide info logging",
```
Comment (FabijanC) [inline starknet_devnet/devnet_config.py:275]: I think it is not clear what we mean by "info logging". Do you mean the server access logging?

File: starknet_devnet/fee_token.py [hunk 15]
```diff
@@ -131,14 +133,26 @@ async def mint(self, to_address: int, amount: int, lite: bool):
...
+        logger.info(transaction)
         starknet: Starknet = self.starknet_wrapper.starknet
         if lite:
             internal_tx = InternalInvokeFunction.from_external(
                 transaction, starknet.state.general_config
             )
-            await starknet.state.execute_tx(internal_tx)
+            execution_info = await starknet.state.execute_tx(internal_tx)
+            logger.info(
+                "transaction execution info: %s", pprint.pformat(execution_info.dump())
```
Comment (FabijanC) [inline starknet_devnet/fee_token.py:144]: Why log only for lite requests?

File: starknet_devnet/starknet_wrapper.py [hunk 16]
```diff
@@ -511,6 +512,14 @@ async def __aexit__(
                     if not self.starknet_wrapper.config.blocks_on_demand:
                         await self.starknet_wrapper.generate_latest_block()
 
+                logger.info(
```
Comment (FabijanC) [inline starknet_devnet/starknet_wrapper.py:515]: I don't understand. Is this always logged?

File: starknet_devnet/starknet_wrapper.py [hunk 17]
```diff
@@ -940,6 +949,9 @@ async def __predeclare_starknet_cli_account(self):
         """Predeclares the account class used by Starknet CLI"""
         state = self.get_state().state
         state.compiled_classes[STARKNET_CLI_ACCOUNT_CLASS_HASH] = oz_account_class
+        if self.config.verbose or not self.config.hide_predeployed_accounts:
```
Comment (FabijanC) [inline starknet_devnet/starknet_wrapper.py:952]: Should the logical operator be `and` or `or`?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: starknet_devnet/devnet_config.py (hunk 14 above)
Comment (tonypony220) [reply starknet_devnet/devnet_config.py:275]: Yes. I decided to add flag and not hide access logs without intention, cause they are useful

Comment (FabijanC) [reply starknet_devnet/devnet_config.py:275]: Ok, can you just change the `help` string then, perhaps by replacing "info logging" with "server access logging"?

File: starknet_devnet/fee_token.py (hunk 15 above)
Comment (tonypony220) [reply starknet_devnet/fee_token.py:144]: regular logs in TransactionHandler

Comment (FabijanC) [reply starknet_devnet/fee_token.py:144]: Then please add a comment line explaining this.

File: starknet_devnet/starknet_wrapper.py (hunk 16 above)
Comment (tonypony220) [reply starknet_devnet/starknet_wrapper.py:515]: yes. Each transaction. But info level logs when verbose only.

File: starknet_devnet/starknet_wrapper.py (hunk 17 above)
Comment (tonypony220) [reply starknet_devnet/starknet_wrapper.py]: `or` because the verbose flag has a higher priority over hiding, meaning it discards all hiding flags. 
It was independent in the first commits, but you pointed out that behavior.

File: page/docs/guide/run.md [hunk 18]
```diff
@@ -7,18 +7,19 @@ sidebar_position: 1
...
                        [--fork-retries FORK_RETRIES] [--chain-id CHAIN_ID] [--disable-rpc-request-validation]
-                       [--disable-rpc-response-validation]
+                       [--disable-rpc-response-validation] [--hide-logs]
 
 Run a local instance of Starknet Devnet
 
 optional arguments:
   -h, --help            show this help message and exit
   -v, --version         Print the version
+  --verbose             Show more verbose output. Have higher priority than --hide-logs and --hide-predeployed-contracts
```
Comment (FabijanC) [inline page/docs/guide/run.md:22]: Have -> Has
This section of `run.md` should be a copy of the output of `starknet-devnet --help`. So any help string you had in mind should first be configured in devnet_config.py

File: starknet_devnet/devnet_config.py [hunk 19]
```diff
@@ -74,6 +74,14 @@ def _fork_block(specifier: str):
     return parsed
 
 
+def _warn_if_used_deprecated_argument(arg: str):
```
Comment (FabijanC) [inline starknet_devnet/devnet_config.py:77]: Leftover if using `WarnIfDeprecatedArgumentAction`?

File: starknet_devnet/devnet_config.py [hunk 20]
```diff
@@ -210,6 +218,22 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+    """
+    Action to warn if user uses old flag;
+    """
+    def __init__(self, nargs=0, **kw):
+        super().__init__(nargs=nargs, **kw)
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        if option_string == "--hide-predeployed-accounts":
+            warn(
+                "WARNING: Argument --hide-predeployed-accounts is deprecated\n",
```
Comment (FabijanC) [inline starknet_devnet/devnet_config.py:231]:
```suggestion
                "WARNING: Argument --hide-predeployed-accounts is deprecated; applying --hide-predeployed-contracts instead",
```
Also, why the `\n`?

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: starknet_devnet/devnet_config.py (hunk 20 above)
Comment (tonypony220) [reply starknet_devnet/devnet_config.py]: To highlight warning. Fixed

File: starknet_devnet/devnet_config.py (hunk 19 above)
Comment (tonypony220) [reply starknet_devnet/devnet_config.py]: My bad

File: starknet_devnet/devnet_config.py (hunk 20 above)
Comment (FabijanC) [reply starknet_devnet/devnet_config.py]: You don't agree with the other part of the suggestion?

File: page/docs/guide/run.md (hunk 18 above)
Comment (FabijanC) [reply page/docs/guide/run.md]: Did you apply this copy-paste rule to all help strings?

**Comment by FabijanC** [comment]
Were you planning to make more changes? Are I just ran your code and noticed the following issues:
- the warning `WARNING: Use these accounts and their keys ...` is logged twice
- chargeable account info is logged - let's not do that, it's an undocumented testing-thing only
- predeclared starknet cli account class hash is logged in decimal instead of hexadecimal format (also, let's write `Starknet CLI` instead of `starknet cli` - it looks better IMO
- thinking about it, the name `--hide-logs` is confusing because every part of output can be considered logging; perhaps rename to `--hide-server-logs` to avoid confusion
- in verbose mode, logging the compressed program brings no benefit IMO; e.g. for my very simple test contract, it logged a the whole ABI pretty-printed and a wall of content like this: `H4sIAKiQiWQC/+y9C5PktpUu+Fc6FLHhx3q68X447myEbPfMOK78WEmzc+/OODJAEFTXuiqrVJklqceh/75gZj0yyWQSJEEQJGGH3d1VJAgcfPjOwes7//hC7...`
